coordinator/src/main/scala/filodb.coordinator/NodeClusterActor.scala

```diff
@@ -0,0 +1,99 @@
+package filodb.coordinator
+
+import akka.actor.{ActorRef, Props, RootActorPath}
+import akka.cluster.Cluster
+import akka.cluster.ClusterEvent._
+import akka.event.LoggingReceive
+import scala.collection.mutable.HashMap
+import scala.concurrent.duration._
+
+import filodb.core.ErrorResponse
+
+object NodeClusterActor {
+  // Forwards message to one random recipient that has the given role.  Any replies go back to originator.
+  case class ForwardToOne(role: String, msg: Any)
+  case object NoSuchRole extends ErrorResponse
+
+  // Gets all the ActorRefs for a specific role.  Returns a Set[ActorRef].
+  case class GetRefs(role: String)
+
+  // Forwards message to all recipients with given role.  Sending actor must handle separate replies.
+  case class ForwardToAll(role: String, msg: Any)
+
+  // Internal message
+  case class AddCoordActor(roles: Set[String], ref: ActorRef)
+
+  def props(cluster: Cluster,
+            resolveActorTimeout: FiniteDuration = 10.seconds): Props =
+    Props(classOf[NodeClusterActor], cluster, resolveActorTimeout)
+}
+
+/**
+ * An actor that subscribes to membership events for a FiloDB cluster, updating a state of coordinators.
+ * It also can send messages to one or all coordinators of a given role.
+ *
+ * Compared to the standard cluster aware routers, it has a couple advantages:
+ * - It tracks all roles, not just one, and can send messages to a specific role per invocation
+ * - It has much lower latency - direct sending to ActorRefs, no need to resolve ActorSelections
+ *   (but of course it assumes the destination actor is always up... TODO watch for other actor dying)
+ * - It broadcasts messages to all members of a role and sends back a collected response in one message,
+ *   making it very easy to handle for non-actors.
+ * - It can notify when some address joins
+ */
+class NodeClusterActor(cluster: Cluster,
+                       resolveActorTimeout: FiniteDuration) extends BaseActor {
+  import NodeClusterActor._
+
+  val roleToCoords = new HashMap[String, Set[ActorRef]]().withDefaultValue(Set.empty[ActorRef])
+
+  import context.dispatcher
+
+  // subscribe to cluster changes, re-subscribe when restart
+  override def preStart(): Unit = {
+    cluster.subscribe(self, initialStateMode = InitialStateAsEvents,
+      classOf[MemberEvent], classOf[UnreachableMember])
+  }
+
+  // TODO: leave cluster?
```

Comment:
Yes, that's all and how/where to leave the cluster, however if the owner of the cluster that is passing it into the constructor here is also an actor, I'd have that actor do it vs here in it's child.